### ISSUE TITLE
Serialization fails for List<CultureInfo>

### ISSUE THREAD

**Issue by DixonDs** [dotnet/orleans#2727]
Here is a failing test:

```
[Fact, TestCategory("BVT"), TestCategory("Functional"), TestCategory("Serialization")]
public void SerializationTests_CultureInfoList()
{
    var input = new List<CultureInfo> { CultureInfo.GetCultureInfo("en"), CultureInfo.GetCultureInfo("de") };

    var output = SerializationManager.RoundTripSerializationForTesting(input);
    Assert.True(input.SequenceEqual(output));
}
```
It gives the following error message: "Named type "CultureInfo" is invalid: Type string "CultureInfo" cannot be resolved."

Serialization of a single `CultureInfo` object works fine though.

**Comment by ReubenBond** [comment]
I could spot-fix this bug, but it's really related to the larger issue of type identity serialization.

Currently, the type serializer only supports types which have been `Register()`d on both ends. Only certain types are registered eagerly. Registering every single type would result in a lot of processing and a very large map of types <-> names.

Solutions include:
* Serializing types using `AssemblyQualifiedName` (but that includes asm versions)
  * ILBasedSerializer uses a method similar to this, but also serializes types using a `TypeKey` struct to improve perf
* Creating a custom version of `AssemblyQualifiedName` which doesn't include asm versions, but is still compatible with `Type.GetType(...)`
* Creating a custom binary representation of type identifiers
  * Assembly identifiers should be serialized at most once per message
  * Generic type parameters will naturally use their most efficient representation (eg, int would be serialized using its `SerializationTokenType` (byte) instead of its fully-qualified type name.)

Also related to #2156, #2597, & #1480

**Comment by ReubenBond** [comment]
Fixed by #3036 - thanks, @DixonDs
